Fix unicode handling in fixup_perms2 errorhandling

noseworthy opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the plugin/module/task -->
action

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
$ ansible --version
ansible 2.2.0.0
  config file = /Users/noseworthy/Development/nocland-devops/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Discovered this issue when trying to modify files on a system as the `postgresql` user. Determined that the output of `stderr` was using smartquotes around the `--become-user` name which were trying to be coerced into ASCII by python 2's "helpful" string handling. Decided that the best way to make sure that this was portable across systems and versions of python would be to ensure that to_text was called on all strings that make up the error message (to avoid use of the `u` prefix on python 2 unicode strings) to ensure that they were in a unicode format (as opposed to an encoded byte string format) so that when `str.format()` was called, no coercion happened and the "unicode sandwich" was in place. 
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

<!-- Paste verbatim command output below, e.g. before and after your change -->
before:
```
$ ansible test_server -m setup --become --become-user=postgresql --ask-become-pass
SUDO password:
An exception occurred during task execution. The full traceback is:
Traceback (most recent call last):
  File "/Users/noseworthy/.virtualenvs/ansible/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 119, in run
    res = self._execute()
  File "/Users/noseworthy/.virtualenvs/ansible/lib/python2.7/site-packages/ansible/executor/task_executor.py", line 490, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/noseworthy/.virtualenvs/ansible/lib/python2.7/site-packages/ansible/plugins/action/normal.py", line 33, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars))
  File "/Users/noseworthy/.virtualenvs/ansible/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 641, in _execute_module
    self._fixup_perms2(remote_files, remote_user)
  File "/Users/noseworthy/.virtualenvs/ansible/lib/python2.7/site-packages/ansible/plugins/action/__init__.py", line 417, in _fixup_perms2
    res['rc'], to_text(res['stderr']))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2018' in position 21: ordinal not in range(128)

test_server | FAILED! => {
    "failed": true,
    "msg": "Unexpected failure during module execution.",
    "stdout": ""
}
```

after:
```
$ ansible test_server -m setup --become --become-user=postgresql --ask-become-pass
SUDO password:
test_server | FAILED! => {
    "failed": true,
    "msg": "Failed to set permissions on the temporary files Ansible needs to create when becoming an unprivileged user (rc: 1, err: chown: invalid user: ‘postgresql’\n). For information on working around this, see https://docs.ansible.com/ansible/become.html#becoming-an-unprivileged-user"
}
```

The _fixup_perms2 method checks to see if the user that is being sudo'd
is an unprivileged user or root. If it is an unprivileged user, some
checks are done to see if becoming this user would lock the ssh user out
of temp files, among other things. If this check fails, an error prints
telling the user to check the documentation for becoming an unprivileged
user.

On some systems, the stderr prints out the unprivileged user the ssh
user was trying to become contained in smartquotes. These quotes aren't
in the ASCII range, and so when we're trying to call `str.format()` to
combine the stderr message with the error text we get a
UnicodeEncodeError as python can't coerce the smartquotes using the
system default encoding. By calling `to_text()` on the error message we
can ensure that the error message is in a python3 friendly unicode
format (`unicode` in py2, `str` in py3) so that no coercion is done and
the python unicode sandwich is maintained.

Fixes: #18444

abadger commented:
A few things need to be changed here:
* Please remove the style changes... Although there is continued debate amongst the core team about style changes, line length restrictions are unlikely to make it in anytime soon.  So those changes would likely (and unfortunately) block the actual bugfix.  Removing the extraneous changes is the quickest route to getting this merged.
* Exception messages need to be native strings (byte strings on python2 and text strings on python3): https://pythonhosted.org/kitchen/unicode-frustrations.html#frustration-5-exceptions  We have a to_native() function that can be used for this  (it's an alias to to_bytes on python2 and to_text on python3).

The code for the latter can look something like this:
``` python
raise AnsibleError('literal strings will be native{0}: {1}'.format(res['rc'], to_native(res['stderr'])))
```

This might provoke an error further up the stack (where the AnsibleError gets caught).  If so, there will need to be a change there as well. to transform the native string into a text string for display.

needs_revision

noseworthy committed:
Fix unicode handling in fixup_perms2 errorhandling

The _fixup_perms2 method checks to see if the user that is being sudo'd
is an unprivileged user or root. If it is an unprivileged user, some
checks are done to see if becoming this user would lock the ssh user out
of temp files, among other things. If this check fails, an error prints
telling the user to check the documentation for becoming an unprivileged
user.

On some systems, the stderr prints out the unprivileged user the ssh
user was trying to become contained in smartquotes. These quotes aren't
in the ASCII range, and so when we're trying to call `str.format()` to
combine the stderr message with the error text we get a
UnicodeEncodeError as python can't coerce the smartquotes using the
system default encoding. By calling `to_native()` on the error message
we can ensure that the error message is a native string for the
`Exception` handling, as `Exception` messages need to be native strings
to avoid errors (byte strings in python2, and text strings in python3)

Fixes: #18444

noseworthy commented:
@abadger, thanks for the review and the comments. Figure'd i'd try to leave the code prettier than I found it, but I understand that "pretty" is in the eye of the beholder and the unending holy war that is code style guides (especially in big projects). I've removed the style stuff and simplified the fix to just call `to_native()` as per your recommendation. Didn't know about the trouble with handling `unicode` messages when handling Exceptions in py2. Thanks for the heads up!

I've run the unittests and did some quick ad-hoc testing and wasn't able to get it to error out higher up the stack as you mentioned. This fix might be sufficient. If there's anything else I can do to help, please let me know :)

noseworthy commented:
Shippable please test this again

noseworthy commented:
Seems to fail on some unrelated `yum` package install?

https://app.shippable.com/runs/58332f9a2bbf381000dfd8c8/8/tests
```
            {
"changed": false, 
"failed": true, 
"msg": "Failure talking to yum: failure: repodata/repomd.xml from rawhide: [Errno 256] No more mirrors to try.\nhttp://download-i2.fedoraproject.org/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.math.princeton.edu/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirror.chpc.utah.edu/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirrors.cat.pdx.edu/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://kdeforge2.unl.edu/mirrors/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.cc.vt.edu/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirrors.lug.mtu.edu/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.cs.pitt.edu/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.cs.princeton.edu/pub/mirrors/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://linux.mirrors.es.net/fedora/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://ftp.linux.ncsu.edu/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://fedora.mirror.lstn.net/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirror.redsox.cc/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.metrocast.net/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirrors.xmission.com/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.web-ster.com/fedora/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.cogentco.com/pub/linux/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirror.steadfast.net/fedora/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirrors.kernel.org/fedora/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.sjc02.svwh.net/fedora/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttp://mirror.prgmr.com/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://dl.fedoraproject.org/pub/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://muug.ca/mirror/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno 12] Timeout on https://muug.ca/mirror/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: (28, 'Connection timed out after 30000 milliseconds')\nhttp://fedora.bhs.mirrors.ovh.net/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirror.its.sfu.ca/mirror/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide\nhttps://mirror.csclub.uwaterloo.ca/fedora/linux/development/rawhide/Everything/x86_64/os/repodata/repomd.xml: [Errno -1] repomd.xml does not match metalink for rawhide"
}
```

abadger commented:
@noseworthy Thanks!  This looks good.  I'll merge it now.  The shippable test failure is unrelated... eventually we'll put small package repositories into our test images.  Until then we'll have transient failures like that.

abadger commented:
Merged to devel and cherry-picked to stable-2.2
